// @airswap/types/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.emptySignature = exports.emptyOrderParty = exports.emptyParty = exports.EIP712Light = exports.EIP712 = void 0;
var constants_1 = require("@airswap/constants");
exports.EIP712 = {
    EIP712Domain: [
        { name: 'name', type: 'string' },
        { name: 'version', type: 'string' },
        { name: 'verifyingContract', type: 'address' },
    ],
    Order: [
        { name: 'nonce', type: 'uint256' },
        { name: 'expiry', type: 'uint256' },
        { name: 'signer', type: 'Party' },
        { name: 'sender', type: 'Party' },
        { name: 'affiliate', type: 'Party' },
    ],
    Party: [
        { name: 'kind', type: 'bytes4' },
        { name: 'wallet', type: 'address' },
        { name: 'token', type: 'address' },
        { name: 'amount', type: 'uint256' },
        { name: 'id', type: 'uint256' },
    ],
};
exports.EIP712Light = {
    EIP712Domain: [
        { name: 'name', type: 'string' },
        { name: 'version', type: 'string' },
        { name: 'chainId', type: 'uint256' },
        { name: 'verifyingContract', type: 'address' },
    ],
    LightOrder: [
        { name: 'nonce', type: 'uint256' },
        { name: 'expiry', type: 'uint256' },
        { name: 'signerWallet', type: 'address' },
        { name: 'signerToken', type: 'address' },
        { name: 'signerAmount', type: 'uint256' },
        { name: 'signerFee', type: 'uint256' },
        { name: 'senderWallet', type: 'address' },
        { name: 'senderToken', type: 'address' },
        { name: 'senderAmount', type: 'uint256' },
    ],
};
exports.emptyParty = {
    kind: constants_1.tokenKinds.ERC20,
    token: constants_1.ADDRESS_ZERO,
    amount: '0',
    id: '0',
};
exports.emptyOrderParty = {
    wallet: constants_1.ADDRESS_ZERO,
    kind: constants_1.tokenKinds.ERC20,
    token: constants_1.ADDRESS_ZERO,
    amount: '0',
    id: '0',
};
exports.emptySignature = {
    version: constants_1.signatureTypes.PERSONAL_SIGN,
    signatory: constants_1.ADDRESS_ZERO,
    validator: constants_1.ADDRESS_ZERO,
    r: constants_1.LOCATOR_ZERO,
    s: constants_1.LOCATOR_ZERO,
    v: '0',
};
