// @airswap/constants/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uniswapRouterAddress = exports.etherscanDomains = exports.balanceCheckerAddresses = exports.stakingTokenAddresses = exports.wethAddresses = exports.signatureTypes = exports.protocolNames = exports.protocols = exports.tokenKindNames = exports.tokenKinds = exports.TokenKinds = exports.chainCurrencies = exports.chainNames = exports.chainIds = exports.SECONDS_IN_DAY = exports.REQUEST_TIMEOUT = exports.DEFAULT_PORT = exports.MIN_CONFIRMATIONS = exports.MAX_APPROVAL_AMOUNT = exports.MAX_LOCATORS = exports.LOCATOR_ZERO = exports.ADDRESS_ZERO = exports.INDEX_HEAD = exports.LIGHT_DOMAIN_VERSION = exports.LIGHT_DOMAIN_NAME = exports.DOMAIN_VERSION = exports.DOMAIN_NAME = void 0;
exports.DOMAIN_NAME = 'SWAP';
exports.DOMAIN_VERSION = '2';
exports.LIGHT_DOMAIN_NAME = 'SWAP_LIGHT';
exports.LIGHT_DOMAIN_VERSION = '3';
exports.INDEX_HEAD = '0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF';
exports.ADDRESS_ZERO = '0x0000000000000000000000000000000000000000';
exports.LOCATOR_ZERO = '0x0000000000000000000000000000000000000000000000000000000000000000';
exports.MAX_LOCATORS = 10;
exports.MAX_APPROVAL_AMOUNT = '90071992547409910000000000';
exports.MIN_CONFIRMATIONS = 2;
exports.DEFAULT_PORT = 3000;
exports.REQUEST_TIMEOUT = 4000;
exports.SECONDS_IN_DAY = 86400;
exports.chainIds = {
    MAINNET: 1,
    RINKEBY: 4,
    GOERLI: 5,
    KOVAN: 42,
    BINANCE: 56,
};
exports.chainNames = {
    1: 'MAINNET',
    4: 'RINKEBY',
    5: 'GOERLI',
    42: 'KOVAN',
    56: 'BINANCE',
};
exports.chainCurrencies = {
    1: 'ETH',
    4: 'Rinkeby ETH',
    5: 'Goerli ETH',
    42: 'Kovan ETH',
    56: 'BNB',
};
var TokenKinds;
(function (TokenKinds) {
    TokenKinds["ERC20"] = "0x36372b07";
    TokenKinds["ERC721"] = "0x80ac58cd";
    TokenKinds["ERC1155"] = "0xd9b67a26";
    TokenKinds["CKITTY"] = "0x9a20483d";
})(TokenKinds = exports.TokenKinds || (exports.TokenKinds = {}));
exports.tokenKinds = {
    ERC20: TokenKinds.ERC20,
    ERC721: TokenKinds.ERC721,
    ERC1155: TokenKinds.ERC1155,
    CKITTY: TokenKinds.CKITTY,
};
exports.tokenKindNames = {
    '0x36372b07': 'ERC20',
    '0x80ac58cd': 'ERC721',
    '0xd9b67a26': 'ERC1155',
    '0x9a20483d': 'CKITTY',
};
exports.protocols = {
    SERVER: '0x0000',
    DELEGATE: '0x0001',
};
exports.protocolNames = {
    '0x0000': 'SERVER',
    '0x0001': 'DELEGATE',
};
exports.signatureTypes = {
    INTENDED_VALIDATOR: '0x00',
    SIGN_TYPED_DATA: '0x01',
    PERSONAL_SIGN: '0x45',
};
exports.wethAddresses = {
    1: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
    4: '0xc778417e063141139fce010982780140aa0cd5ab',
    5: '0xb4fbf271143f4fbf7b91a5ded31805e42b2208d6',
    42: '0xd0a1e359811322d97991e03f863a0c30c2cf029c',
};
exports.stakingTokenAddresses = {
    1: '0x27054b13b1b798b345b591a4d22e6562d47ea75a',
    4: '0xcc1cbd4f67cceb7c001bd4adf98451237a193ff8',
    5: '0x1a1ec25dc08e98e5e93f1104b5e5cdd298707d31',
    42: '0x1a1ec25dc08e98e5e93f1104b5e5cdd298707d31',
    56: '0x1ac0d76f11875317f8a7d791db94cdd82bd02bd1',
};
exports.balanceCheckerAddresses = {
    1: '0x5dfe850d4b029c25c7ef9531ec9986c97d90300f',
    4: '0xa1e2c4132cbd33c3876e1254143a850466c97e32',
    5: '0x755aa03f420a62560e90502d7da23a73c301dad4',
    42: '0xe25b7504856bfb230b7c32bc87047479815cbc70',
};
exports.etherscanDomains = {
    1: 'etherscan.io',
    4: 'rinkeby.etherscan.io',
    5: 'goerli.etherscan.io',
    42: 'kovan.etherscan.io',
    56: 'bscscan.com',
};
exports.uniswapRouterAddress = '0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D';
